### PR TITLE
Add nondeterministic alert to index_copy, median CUDA and kthvalue CUDA

### LINKED ISSUE #46743: Add nondeterministic error for `index_copy_` if duplicate indices are given
`index_copy_` is nondeterministic if the indices argument contains duplicate entries. When `torch.set_deterministic(True)` is set, it should produce an error.

### PR DESCRIPTION
**Pull request by kurtamohler**
Also fixes issue where skipped tests did not properly restore deterministic flag.

Fixes #46743

### CODE REVIEW

File: aten/src/ATen/cuda/LegacyTHFunctionsCUDA.cpp [hunk 1]
```diff
@@ -362,6 +363,22 @@ Tensor & _th_masked_scatter_bool_(Tensor & self, const Tensor & mask, const Tens
     }
     return self;
 }
+
+static void alert_nondeterministic_duplicate_indices(const Tensor& indices) {
```
Comment (kurtamohler) [inline aten/src/ATen/cuda/LegacyTHFunctionsCUDA.cpp:367]: Note, it would be better to do this check while the indices are being accessed in the implementation of index_copy (like I'm doing for the CPU version), to avoid this extra traversal. However, I don't think there's a very practical way to do that, considering that the CUDA implementation is split between gpu threads. Writing a duplication check algorithm that works across threads would be possible, but challenging.

**Comment by ngimel** [comment]
Quick notes (did not do a full review)
1) you are changing generated file, this is not good. 
2) for scatter_add, we warn regardless of whether there are repeating indices, so maybe we should do this for index_copy too https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/ScatterGatherKernel.cu#L453-L460
3) while I was looking at scatter_add, I noticed that scatter is not warning, and yet for repeating indices scatter produces nondeterministic result.

File: aten/src/ATen/cuda/LegacyTHFunctionsCUDA.cpp (hunk 1 above)
Comment (mruberry) [reply aten/src/ATen/cuda/LegacyTHFunctionsCUDA.cpp:367]: @ngimel and I recommend that, like other operations, there not be a check for duplicate indices (at least for now) and, instead, these operations are just labeled nondeterministic and prohibited

File: aten/src/TH/generic/THTensorEvenMoreMath.cpp [hunk 2]
```diff
@@ -168,6 +169,12 @@ void THTensor_(indexCopy)(THTensor *tensor, int dim, THLongTensor *index, THTens
 
   index = THLongTensor_newContiguous(index);
   index_data = THLongTensor_data(index);
+  std::unordered_set<int64_t> visited_indices;
+  bool is_deterministic = at::globalContext().deterministic();
+  char duplicate_error_msg[] = (
+      "index_copy acts nondeterministically with duplicate indices, "
```
Comment (mruberry) [inline aten/src/TH/generic/THTensorEvenMoreMath.cpp:175]: This message will need an update if the check isn't performed.

File: aten/src/TH/generic/THTensorEvenMoreMath.cpp [hunk 3]
```diff
@@ -176,7 +183,13 @@ void THTensor_(indexCopy)(THTensor *tensor, int dim, THLongTensor *index, THTens
 
     for (i=0; i<numel; i++)
     {
-      THTensor_(select)(tSlice, tensor, dim, index_data[i]);
+      int64_t index_val = index_data[i];
```
Comment (mruberry) [inline aten/src/TH/generic/THTensorEvenMoreMath.cpp:186]: Per the above, let's not check for duplicates and always error (at least for now)

File: test/test_torch.py [hunk 4]
```diff
@@ -12666,6 +12666,20 @@ def test_index_copy(self, device):
         c = torch.zeros(3)
         self.assertRaises(IndexError, lambda: a.index_copy_(dim=1, index=torch.tensor([3]), source=c))
 
+    # Ensure that index_copy throws an error if torch.set_deterministic(True)
+    # is set and duplicate indices are given
+    @wrapDeterministicFlagAPITest
+    def test_index_copy_nondeterministic(self, device):
+        a = torch.randn(100, 4, device=device)
```
Comment (mruberry) [inline test/test_torch.py:12673]: This test will need a small update where indices are not duplicative.

**Comment by mruberry** [comment]
Hey @kurtamohler! Great to see more nondeterministic behavior caught by this flag. For now @ngimel and I think we should just always error like we do with other operations. It's an interesting design question to follow-up with for whether we should add these checks to all operations that are nondeterministic when given duplicate indices.

**Comment by mruberry** [comment]
Follow-up: we should also throw an error if the determinism flag is set and indices are returned for median or kth value on CUDA.

**Comment by kurtamohler** [comment]
Tests for `adaptive_log_softmax` are failing because they depend on `index_copy`, and they are evidently setting `torch.set_deterministic(True)` somewhere. I'm not yet sure what is setting it.

**Comment by kurtamohler** [comment]
Looks like the `expectedAlertNondeterministic` test decorator does not work properly for tests that are skipped. In this case, `torch.set_deterministic(True)` is set, and the initial state is never recovered. This is the cause of the `adaptive_log_softmax` failures. I'll fix the decorator.

**Comment by kurtamohler** [comment]
Barring any additional test failures, I think this done.

File: aten/src/ATen/native/cuda/Sorting.cu [hunk 5]
```diff
@@ -316,6 +316,9 @@ std::tuple<Tensor&, Tensor&> median_with_indices_impl(
     int64_t dim,
     bool keepdim,
     bool ignore_nan) {
+  // See note [Writing Nondeterministic Operations]
+  // Nondeterministic if indices output contains duplicates
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/Sorting.cu:320]: More accurately, it's if there are multiple possible median elements that the indices returned are nondeterminism. This is probably something we should fix, by the way. We fixed argmin and argmax, for example, to always return the indices of the first minimum or maximum element.

Comment (mruberry) [reply aten/src/ATen/native/cuda/Sorting.cu:320]: I'm happy to see the docs actually explain this: 

indices does not necessarily contain the first occurrence of each median value found, unless it is unique. The exact implementation details are device-specific. Do not expect the same result when run on CPU and GPU in general. For the same reason do not expect the gradients to be deterministic.

File: aten/src/ATen/native/cuda/Sorting.cu [hunk 6]
```diff
@@ -410,6 +413,9 @@ std::tuple<Tensor&, Tensor&> kthvalue_out_cuda(
     int64_t k,
     int64_t dim,
     bool keepdim) {
+  // See note [Writing Nondeterministic Operations]
+  // Nondeterministic if indices output contains duplicates
+  at::globalContext().alertNotDeterministic("kthvalue CUDA with indices output");
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/Sorting.cu:418]: Unfortunately, unlike median, the kthvalue documentation does not mention its possible nondeterminism. I think the nondeterminism occurs if there are multiple valid kth values, in which case any index can be returned. This PR should probably update the documentation to reflect this, too.

Is it really only kth value on CUDA that has this issue? If so, we should be explicit about which indices the CPU version will return in the documentation.

**Comment by mruberry** [comment]
Hey @kurtamohler, made a few inline suggestions to align the comments and documentation.

Is there a good determinism test that verifies this runtime error will trigger when these functions are called? It'd be good to see it triggered on the function, method, inplace, and out variants. For median it should not be triggered if indices aren't returned but triggered if they are.

File: aten/src/ATen/native/cuda/Sorting.cu (hunk 6 above)
Comment (kurtamohler) [reply aten/src/ATen/native/cuda/Sorting.cu:418]: Thanks for the extra details! Do you know if there is an issue somewhere describing the nondeterminism of `kthvalue`? I only put the alert in the CUDA version because in a comment above, you said "kth value on CUDA."

Comment (mruberry) [reply aten/src/ATen/native/cuda/Sorting.cu:418]: I don't know if there's an issue for kthvalue, sorry. We could always file an extra one and resolve it as a duplicate later if it is redundant.

That's true, and it probably is just CUDA, but it might be worth checking and documenting the CPU behavior.

Comment (kurtamohler) [reply aten/src/ATen/native/cuda/Sorting.cu:418]: Ok I understand `kthvalue`'s nondeterminism more clearly now after writing a test: https://github.com/kurtamohler/pytorch-perf-test-scripts/blob/master/determinism_test/kthvalue_determinism.py

In the cases I tried, CPU looks like it might be deterministic, but CUDA is not, depending on the inputs. Results are below.

Aside from determinism, there's another interesting aspect--if the input has N duplicate elements, do we get a unique index output from `kthvalue` for each corresponding k argument?

For example, with an input tensor `a = torch.tensor([0.12, 3.4, 0.12, 5.2])`, we have duplicate 0.12 elements at index 0 and 2. The k arguments for these are 1 and 2, since 0.12 will be and indices 0 and 1 after sorting the tensor. The question is, when we call `a.kthvalue(1)` and `a.kthvalue(2)`, do the outputs of those call give two different index values, 0 and 2? If so, then we can say that the "k to index mapping" is unique, because each k value will give a different index. CPU gave me unique mappings for each case I tried. CUDA never gave me unique mappings for duplicates, but for some inputs the mapping was at least consistent--the returned index value was always the same, for each of the duplicates' k values. For an example of a consistent but not unique mapping, in the above simple example, imagine that `a.kthvalue(1)` and `a.kthvalue(2)` both return an index of 0.

<details>

<summary> Click for results </summary>


```
device size number-of-copies k-to-index-unique-mapping maybe-deterministic
cpu 20 1 True True
cpu 20 2 True True
cpu 20 5 True True
cpu 20 10 True True
cpu 20 11 True True
cpu 100 1 True True
cpu 100 2 True True
cpu 100 5 True True
cpu 100 10 True True
cpu 100 11 True True
cpu 1000 1 True True
cpu 1000 2 True True
cpu 1000 5 True True
cpu 1000 10 True True
cpu 1000 11 True True
cpu 10000 1 True True
cpu 10000 2 True True
cpu 10000 5 True True
cpu 10000 10 True True
cpu 10000 11 True True
cuda 20 1 True True
cuda 20 2 False True
cuda 20 5 False True
cuda 20 10 False True
cuda 20 11 False True
cuda 100 1 True True
cuda 100 2 False True
cuda 100 5 False True
cuda 100 10 False True
cuda 100 11 False True
cuda 1000 1 True True
cuda 1000 2 N/A False
cuda 1000 5 N/A False
cuda 1000 10 N/A False
cuda 1000 11 N/A False
cuda 10000 1 True True
cuda 10000 2 N/A False
cuda 10000 5 N/A False
cuda 10000 10 N/A False
cuda 10000 11 N/A False
```

</details>

I have some questions about what to do next:
* My tests do not show that CPU is nondeterministic, but we can't be sure that it isn't. Should we just assume that CPU is nondeterministic and add the alert, or should we leave it alone until someone finds a nondeterministic case?
* I can document what I know about CUDA's nondeterminism. What about the question of unique k-to-index mappings? Is that important or useful for users to know?

Comment (ngimel) [reply aten/src/ATen/native/cuda/Sorting.cu:418]: Please check cpu implementation for the sources of nondeterminism, it's a question to which we can know the exact answer. 
As for cuda, there are no guarantees of consistent mapping, any of the indices corresponding to a particular value can be returned. It may appear deterministic on some hardware/small examples, but don't be deceived by it.

Comment (kurtamohler) [reply aten/src/ATen/native/cuda/Sorting.cu:418]: I read through the kthvalue CPU code, and I don't see anything that would be an obvious source of nondeterminism, as far as I can tell. There is a parallel aspect of kthvalue CPU, but it's just for splitting up the separate arrays that need to be sorted (which happens if the input tensor has more than one dimension), so it is not a source of nondeterminism. On each array, a serial quicksort algorithm is used and then the kth value is pulled from the sorted array. The quicksort looks deterministic to me: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Sorting.cpp#L36

**Comment by kurtamohler** [comment]
I made some documentation updates and wrote some more robust tests for these three operations to ensure they do and don't throw errors in the right cases.

**Comment by mruberry** [comment]
> I made some documentation updates and wrote some more robust tests for these three operations to ensure they do and don't throw errors in the right cases.

Sounds great - is this ready for review?

**Comment by kurtamohler** [comment]
> Sounds great - is this ready for review?

Almost--I need to find out how to fix the buffer overflow one of my tests is causing in `pytorch_linux_xenial_py3_clang5_asan_test1`

**Comment by kurtamohler** [comment]
I found out that the buffer overflow happens when `kthvalue`'s first `out` argument is the same as the input tensor, like `torch.kthvalue(a, k, out=(a, indices))`. I had added a test case that does this, so I decided to just disable that test case for now. I captured a log of the failure, so I can submit an issue to fix this overflow.

I think this is ready for another review when you get a chance @mruberry

**Comment by ngimel** [comment]
I think we should disallow overlapping outputs in this case, kthvalue at least on cuda won't produce correct result.

**Comment by kurtamohler** [comment]
Ah ok,makes sense. I can add an overlap check and error in a different PR.

File: torch/_torch_docs.py [hunk 7]
```diff
@@ -3824,6 +3824,13 @@ def merge_dicts(*dicts):
 (see :func:`torch.squeeze`), resulting in both the :attr:`values` and
 :attr:`indices` tensors having 1 fewer dimension than the :attr:`input` tensor.
 
+.. warning::
```
Comment (mruberry) [inline torch/_torch_docs.py:3827]: Let's turn this warning into a note and simplify it to:

"When there are multiple valid kth values this function may nondeterministically return indices for any of them."

File: torch/testing/_internal/common_utils.py [hunk 8]
```diff
@@ -398,6 +398,18 @@ def wrapper(*args, **kwargs):
             fn(*args, **kwargs)
     return wrapper
 
+# RAII guard for setting deterministic flag and automatically resetting it
+# to its original value
+class DeterministicGuard:
+    def __init__(self, deterministic=None):
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:404]: I suggest removing the default value for deterministic and requiring the caller always specify whether they want deterministic on or off for readability.

File: torch/testing/_internal/common_utils.py [hunk 9]
```diff
@@ -398,6 +398,18 @@ def wrapper(*args, **kwargs):
             fn(*args, **kwargs)
     return wrapper
 
+# RAII guard for setting deterministic flag and automatically resetting it
+# to its original value
+class DeterministicGuard:
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:403]: I was expecting this to be modeled as a context manager with dunder enter and dunder exit. What would you think of that to be more Pythonic?

File: torch/testing/_internal/common_utils.py [hunk 10]
```diff
@@ -408,33 +420,29 @@ def wrapper(*args, **kwargs):
...
-                    os.environ[cublas_var_name] = cublas_config_restore
-        try:
-            fn(*args, **kwargs)
-        except RuntimeError:
-            restore()
-            raise
-        else:
-            restore()
+
+            def __init__(self):
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:428]: Again I don't think we want to model a decorator as using dunder init and dunder del. There are some examples of different kinds of Pythonic decorators in https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_device_type.py that may be interesting to look at.

**Review by mruberry — COMMENTED** [review]
Hey @kurtamohler!

Overall the actual change looks straightforward and good. Just a few questions/comments about "Python-ness" of the testing stuff.

File: torch/testing/_internal/common_utils.py (hunk 9 above)
Comment (kurtamohler) [reply torch/testing/_internal/common_utils.py:403]: Makes sense, I've updated it

File: torch/testing/_internal/common_utils.py (hunk 10 above)
Comment (kurtamohler) [reply torch/testing/_internal/common_utils.py]: I've changed `CuBLASConfigGuard` to act like a context manager

**Comment by kurtamohler** [comment]
I'm not sure what is causing the `pytorch-linux-bionic-rocm3.9-py3.6` failures. They are all due to incorrect values returned by `remainder` CUDA with float16 and bfloat16. It doesn't seem that my changes would cause this, but I don't see any upstream failures in the CI HUD. Are these flaky tests?

**Comment by ngimel** [comment]
There was upstream failure for this, rebase and you should be fine.

**Comment by kurtamohler** [comment]
Had to rebase to fix a conflict. I think this is ready to go.

File: test/test_torch.py [hunk 11]
```diff
@@ -3469,6 +3490,30 @@ def run_test(x, y):
         y[1] = 1.
         run_test(x, y)
 
+    # Ensures that median throws nondeterministic alerts in the correct cases
+    @dtypes(torch.double)
+    def test_median_nondeterministic_alert(self, device, dtype):
+        def test_func(slf, device, call_type):
```
Comment (mruberry) [inline test/test_torch.py:3496]: This doesn't appear to be called anywhere?

File: torch/_torch_docs.py [hunk 12]
```diff
@@ -3894,6 +3894,10 @@ def merge_dicts(*dicts):
 (see :func:`torch.squeeze`), resulting in both the :attr:`values` and
 :attr:`indices` tensors having 1 fewer dimension than the :attr:`input` tensor.
 
+.. note::
+    When there are multiple valid :attr:`k` th values, this function may
```
Comment (mruberry) [inline torch/_torch_docs.py:3898]: Only if the input is CUDA, though. 

"When :attr:\`input\` is on CUDA and there are multiple valid :attr:\`k\`th values, this function may nondeterministically return :attr:\`indices\` for any of them."

File: torch/_torch_docs.py [hunk 13]
```diff
@@ -3894,6 +3894,10 @@ def merge_dicts(*dicts):
 (see :func:`torch.squeeze`), resulting in both the :attr:`values` and
 :attr:`indices` tensors having 1 fewer dimension than the :attr:`input` tensor.
 
+.. note::
```
Comment (mruberry) [inline torch/_torch_docs.py:3897]: Index_copy_ needs a note, too:

https://pytorch.org/docs/master/tensors.html?highlight=index_copy#torch.Tensor.index_copy_

but median already has one:

https://pytorch.org/docs/master/generated/torch.median.html?highlight=median#torch.median

**Comment by mruberry** [comment]
Hey @kurtamohler! Had a look, just a few more comments/suggestions.

File: torch/testing/_internal/common_utils.py [hunk 14]
```diff
@@ -408,33 +421,28 @@ def wrapper(*args, **kwargs):
 def wrapDeterministicFlagAPITest(fn):
```
Comment (mruberry) [inline torch/testing/_internal/common_utils.py:421]: Please elaborate in this comment that tests using this wrapper need to start a subprocess (how and why) so that their cuBLAS is initialized with the new workspace config

File: test/test_torch.py (hunk 11 above)
Comment (kurtamohler) [reply test/test_torch.py:3496]: Whoops, I must have accidentally chopped off the calls when I rebased. I'll add them back in

File: torch/testing/_internal/common_utils.py (hunk 14 above)
Comment (kurtamohler) [reply torch/testing/_internal/common_utils.py:443]: Added, let me know what you think

File: torch/_torch_docs.py (hunk 13 above)
Comment (kurtamohler) [reply torch/_torch_docs.py:3897]: Done

**Review by mruberry — APPROVED** [review]
Cool! Thanks Kurt!
